harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -240,106 +518,281 @@ def on_shutdown(callback_name: str, on_trial_shutdown: Callable) -> None:
             #
             # We create it before loading state because we don't want the training_iterator
             # shuffling values after we load state.
+
+            # XXX: remove training_iterator
             self.training_iterator = iter(self.training_loader)
+            self.training_enumerator = enumerate(self.training_iterator)
 
             def cleanup_iterator() -> None:
                 # Explicitly trigger the training iterator's shutdown (which happens in __del__).
                 # See the rather long note in pytorch/torch/utils/data/dataloader.py.
                 del self.training_iterator
+                del self.training_enumerator
 
             exit_stack.enter_context(defer(cleanup_iterator))
 
             # If a load path is provided load weights and restore the data location.
-            if self.env.latest_checkpoint is not None:
-                logging.info(f"Restoring trial from checkpoint {self.env.latest_checkpoint}")
-                with self.context._core.checkpoint.restore_path(
-                    self.env.latest_checkpoint
+            if self._latest_checkpoint is not None:
+                logging.info(f"Restoring trial from checkpoint {self._latest_checkpoint}")
+                with self._context._core.checkpoint.restore_path(
+                    self._latest_checkpoint
                 ) as load_path:
                     self._load(load_path)
 
-            if self.context.distributed.size > 1 and self.use_horovod:
-                hvd.broadcast_parameters(self.context._main_model.state_dict(), root_rank=0)
-                for optimizer in self.context.optimizers:
+            if self._context.distributed.size > 1 and self._use_horovod:
+                hvd.broadcast_parameters(self._context._main_model.state_dict(), root_rank=0)
+                for optimizer in self._context.optimizers:
                     hvd.broadcast_optimizer_state(optimizer, root_rank=0)
 
-            with self.prof:
-                for callback in self.callbacks.values():
-                    with self.prof.record_timing(
-                        f"callbacks.{callback.__class__.__name__}.on_training_start"
-                    ):
-                        callback.on_training_start()
-                self._run()
-
-    def _run(self) -> None:
```

Comment:
I still really think this code will be much more approachable if you separate the setup code from the iterating-through-ops code.  I still think `_run()` is a useful breakpoint between the two functionalities.

Reply:
i agree with that. changed.